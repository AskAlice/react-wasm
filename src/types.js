// @flow

import type { Node } from 'react';

export type ResultObject = {
  module: WebAssembly.Module,
  instance: WebAssembly.Instance
};

type RenderProps = {
  loading: boolean,
  error: ?any,
  data: ?ResultObject
};

export type WasmParams = {
  url?: ?string,
  bufferSource?: ?BufferSource,
  importObject?: ?{}
};

export type WasmProps = {
  ...$Exact<WasmParams>,
  children: (RenderProps: RenderProps) => Node
};

export type WasmState = RenderProps;
