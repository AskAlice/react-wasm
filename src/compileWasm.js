// @flow

import type { WasmParams, ResultObject } from './types';

const compileWasm = ({
  url,
  bufferSource,
  importObject
}: WasmParams): Promise<ResultObject> =>
  Promise.resolve().then(() => {
    let res;

    if (typeof url === 'string') {
      res = fetch(url);
      if (WebAssembly.instantiateStreaming !== undefined) {
        return WebAssembly.instantiateStreaming(res, importObject);
      }
      res = res.then(response => response.arrayBuffer());
    } else if (bufferSource) {
      res = Promise.resolve(bufferSource);
    } else {
      throw new Error(
        'Can\'t instantiate WebAssembly module, invalid parameters.'
      );
    }

    return res
      .then(buff => WebAssembly.compile(buff))
      .then(module =>
        WebAssembly.instantiate(module, importObject).then(instance => ({
          module,
          instance
        }))
      );
  });

export default compileWasm;
