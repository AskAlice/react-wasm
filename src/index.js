// @flow

import React from 'react';
import withWasmDefinition from './withWasm';
import compileWasm from './compileWasm';
import type { WasmProps, WasmState } from './types';

export default class Wasm extends React.Component<WasmProps, WasmState> {
  static defaultProps = {
    url: null,
    bufferSource: null,
    importObject: {}
  };

  state = {
    loading: true,
    error: null,
    data: null
  };

  componentDidMount() {
    const { url, bufferSource, importObject } = this.props;

    compileWasm({
      url,
      bufferSource,
      importObject
    })
      .then(({ module, instance }) => {
        this.setState({
          loading: false,
          error: null,
          data: {
            module,
            instance
          }
        });
      })
      .catch(error => {
        this.setState({
          loading: false,
          error,
          data: null
        });
      });
  }

  render() {
    const { children } = this.props;

    return children(this.state);
  }
}

export const withWasm = withWasmDefinition;
