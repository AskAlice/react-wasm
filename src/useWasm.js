// @flow

import { useState, useEffect } from 'react';
import compileWasm from './compileWasm';
import type { WasmParams } from './types';

const useWasm = ({
  url,
  bufferSource,
  importObject
}: WasmParams) => {
  const [state, setState] = useState({
    loading: true,
    error: null,
    data: null
  });

  const [prevProps, setPrevProps] = useState({
    url: null,
    bufferSource: null
  });

  let newState = state;

  if (
    (url && url !== prevProps.url) ||
    (!url && bufferSource !== prevProps.bufferSource)
  ) {
    newState = {
      loading: true,
      error: null,
      data: null
    };

    setState(newState);
    setPrevProps({
      url,
      bufferSource
    });
  }

  useEffect(() => {
    compileWasm({
      url,
      bufferSource,
      importObject
    })
      .then(({ module, instance }) => {
        setState({
          loading: false,
          error: null,
          data: {
            module,
            instance
          }
        });
      })
      .catch(ex => {
        setState({
          loading: false,
          error: ex,
          data: null
        });
      });
  }, [
    url,
    !url && bufferSource,
    importObject,
    setState
  ]);

  return newState;
}

export default useWasm;
