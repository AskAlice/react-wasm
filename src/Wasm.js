// @flow

import useWasm from './useWasm';
import type { WasmProps } from './types';

const Wasm = ({ url, bufferSource, importObject, children }: WasmProps) => {
  const state = useWasm({
    url,
    bufferSource,
    importObject
  });

  return children(state);
};

Wasm.defaultProps = {
  url: null,
  bufferSource: null,
  importObject: {}
};

export default Wasm;
