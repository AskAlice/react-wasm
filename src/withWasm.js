import Wasm from '.';

export const withWasm = ComponentDefinition => null;
