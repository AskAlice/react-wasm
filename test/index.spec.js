import Wasm, { useWasm, withWasm } from '../src/index';
/* eslint-disable */
import { default as useWasmHook } from '../src/useWasm';
import { default as WasmComponent } from '../src/Wasm';
import { default as withWasmHoC } from '../src/withWasm';
/* eslint-enable */

describe('react-wasm', () => {
  it('should export useWasm', () => {
    expect(useWasm).toBeTruthy();
    expect(useWasm).toEqual(useWasmHook);
  });

  it('should export Wasm', () => {
    expect(Wasm).toBeTruthy();
    expect(Wasm).toEqual(WasmComponent);
  });

  it('should export withWasm', () => {
    expect(withWasm).toBeTruthy();
    expect(withWasm).toEqual(withWasmHoC);
  });
});
