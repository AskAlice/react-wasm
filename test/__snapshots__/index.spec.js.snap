// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-wasm should load module from ArrayBuffer 1`] = `
<div>
  loading: 
  false
  error: 
  null
  1 + 2 = 
  3
  20 / 2 = 
  10
</div>
`;

exports[`react-wasm should load module from ArrayBuffer with importObject 1`] = `
<div>
  loading: 
  false
  error: 
  null
  1 + 2 = 
  3
</div>
`;

exports[`react-wasm should load module from url 1`] = `
<div>
  loading: 
  false
  error: 
  null
  1 + 2 = 
  3
  20 / 2 = 
  10
</div>
`;

exports[`react-wasm should load module from url with importObject 1`] = `
<div>
  loading: 
  false
  error: 
  null
  1 + 2 = 
  3
</div>
`;

exports[`react-wasm should set error if bufferSource is provided with wrong type 1`] = `"{\\"loading\\":false,\\"error\\":{},\\"data\\":null}"`;

exports[`react-wasm should set error if invalid ArrayBuffer is provided 1`] = `"{\\"loading\\":false,\\"error\\":{},\\"data\\":null}"`;

exports[`react-wasm should set error if invalid url is provided 1`] = `"{\\"loading\\":false,\\"error\\":{},\\"data\\":null}"`;

exports[`react-wasm should set error if no url and bufferSource are provided 1`] = `"{\\"loading\\":false,\\"error\\":{},\\"data\\":null}"`;

exports[`react-wasm should set error if url is provided with wrong type 1`] = `"{\\"loading\\":false,\\"error\\":{},\\"data\\":null}"`;

exports[`react-wasm should set loading 1`] = `"{\\"loading\\":true,\\"error\\":null,\\"data\\":null}"`;
